src/main/kotlin/com/github/treepat/match_manager/MatchManager.kt

```diff
@@ -0,0 +1,64 @@
+package com.github.treepat.match_manager
+
+import com.github.treepat.expression.TreepatExpression
+import com.github.treepat.target_tree.TargetTree
+import com.github.treepat.target_tree.TargetTreeNode
+
+class MatchManager(private var expression: TreepatExpression, private var tree: TargetTree) {
+
+    private var nextNodeToEvaluate: TargetTreeNode? = null
+
+    init {
+        this.nextNodeToEvaluate = tree.root
+    }
+
+    fun getNextMatch(): List<TargetTreeNode>? {
+        if (nextNodeToEvaluate == null) {
+            nextNodeToEvaluate = tree.root
+            return null
+        }
+
+        var response = expression.executeExpression(nextNodeToEvaluate)
+        this.nextNodeToEvaluate = nextNodeToEvaluate!!.nextPreorderNode()
+
+        while ((!response.hasMatch || response.responses.filter { it.matches.isNotEmpty() }.isEmpty()) && nextNodeToEvaluate != null) {
+            response = expression.executeExpression(nextNodeToEvaluate)
+            nextNodeToEvaluate = nextNodeToEvaluate!!.nextPreorderNode()
+        }
+
+        var foundResponse: List<TargetTreeNode>? = null
+
+        if (response.hasMatch && response.responses.filter { it.matches.isNotEmpty() }.isNotEmpty()) {
+            foundResponse = response.responses.maxBy { it.matches.size }?.matches
+        }
+        return foundResponse
+    }
```

Comment:
`getNextMatch()` Gives ONLY the next match. When it reaches the end of the tree and doesn't find another match it returns `null`. After returning `null` `MatchManager` resets and starts over.

Useful if you want to work on matches one by one.

Reply:
This comment would be useful to add as a JavaDoc in the code.